app/javascript/mastodon/features/interaction_modal/index.jsx

```diff
@@ -18,62 +25,207 @@ const mapDispatchToProps = (dispatch) => ({
   },
 });
 
-class Copypaste extends React.PureComponent {
+const PERSISTENCE_KEY = 'mastodon_home';
+
+const isValidDomain = value => {
+  const url = new URL('https:///path');
+  url.hostname = value;
+  return url.hostname === value;
+};
+
+const valueToDomain = value => {
+  // If the user uses a URL to their profile page or server
+  if (/^https?:\/\//.test(value)) {
+    try {
+      return (new URL(value)).host;
+    } catch {
+      return '';
+    }
+  // If the user writes their full handle including username
+  } else if (value.includes('@')) {
+    const segments = value.split('@');
+
+    if (segments.length === 2) {
+      return segments[1];
+    } else {
+      return '';
+    }
+  }
+
+  return value;
+};
+
+class LoginForm extends React.PureComponent {
 
   static propTypes = {
-    value: PropTypes.string,
+    resourceUrl: PropTypes.string,
+    intl: PropTypes.object.isRequired,
   };
 
   state = {
-    copied: false,
+    value: localStorage ? (localStorage.getItem(PERSISTENCE_KEY) || '') : '',
```

Comment:
What happens if the stored value is corrupted? Should we validate it after loading from localstorage?